### PR TITLE
Stack API spec

### LINKED ISSUE #57: Stack management API
- [ ] Deploy stack given name and stack file contents - `docker stack deploy -c <file_name> <stack_name>`
- [ ] Get list of running stacks - `docker stack ls`
- [ ] Remove running stack - `docker stack rm <stack_name>`
- [ ] Get list of running services in a stack* - `docker stack services <stack_name>`
- [ ] Get list of tasks running in a stack* - `docker stack ps <stack_name>`

\* Page has not yet been planned to use functionality. These should be worked on as last priority

### Considerations
- Format of stack file contents
- Format of returned data

### PR DESCRIPTION
**Pull request by egeldenhuys**
Closes #57 

The rendered spec can be viewed [here](https://tripleparity.github.io/docks-api/stack-api-spec.html)

Rendered using
```
npm install -g aglio
aglio -i docks.apib --theme-template triple -o stack-api-spec.html
```

Themes can be viewed in the [Aglio Repository](https://github.com/danielgtaylor/aglio#example-output)

### CODE REVIEW

**Comment by egeldenhuys** [comment]
> I'm still not sure how we will implement the update functionality but we'll get to it.

The most complex part will be to reverse the docker inspect, which will also require discovering all services related to a stack. From there we can use `docker stack deploy` which will update the stack, given a docker-compose file

**Comment by egeldenhuys** [comment]
I realise the endpoint for retrieving a reversed `docker-compose.yml` for a stack is missing.

Rough Draft:

```apib
### Get Docker Compose File [GET /stacks/{stackName}/stackfile]

+ Paramaters
    + stackName (string) - Unique stack name

+ Response 200 (application/json)
        {
            "data": {
                "stackFile": "IyBEZXBsb3kgdG...V09SRD1leGFtcGxl"
            }
        }

+ Response 404
    Stack does not exist

    + Body

```

The reason for having `GET /stacks/{stackName}/stackfile` instead of `GET /stacks/{stackName` is that generating the stack file will be an expensive operation and should only be used when required.

The `Deploy New Stack` endpoint documentation looks strange, I will also fix that,

**Comment by annamarieHelberg** [comment]
This looks amazing! Thank you for implementing this!

**Comment by egeldenhuys** [comment]
@annamarieHelberg Note that this is only the contract between front-end and back-end, the real work still has to be done :laughing:

File: docks.apib [hunk 1]
```diff
@@ -93,35 +91,182 @@ Create a new user
...
+        + (object)
+            + stackName (string, required) - The unique name for the stack
+            + servicesCount (number, required) - The number of services running in the stack
+            
++ Response 200 (application/json)
+
+        {
+            "data": [
+                {
+                    "name": "docks-api",
```
Comment (egeldenhuys) [inline docks.apib]: Note that this field should be `stackName` and not `name`

**Comment by egeldenhuys** [comment]
When returning tasks we also need to conform to the existing Docker API data structure

TODO:
- [ ] Return Services according to Docker API
- [ ] Return Tasks according to Docker API
- [ ] Add endpoint to get stack file

**Comment by egeldenhuys** [comment]
The schema for `/stacks/{stackName}/services` will follow
https://docs.docker.com/engine/api/v1.37/#operation/ServiceList
with the addition of being wrapped in `"data": [<SERVICES>]`

The schema for `/stacks/{stackName}/tasks` will follow
https://docs.docker.com/engine/api/v1.37/#operation/TaskList
with the addition of being wrapped in `"data": [<TASKS>]`

I do not see a good reason for duplicating the data structure in the documentation.
If we were using the same specification format (Swagger) as Docker API, this would be a quick fix

**Comment by egeldenhuys** [comment]
The schema for `/stacks/{stackName}/services` will follow
https://docs.docker.com/engine/api/v1.37/#operation/ServiceList
with the addition of being wrapped in `"data": [<SERVICES>]`

The schema for `/stacks/{stackName}/tasks` will follow
https://docs.docker.com/engine/api/v1.37/#operation/TaskList
with the addition of being wrapped in `"data": [<TASKS>]`

I do not see a good reason for duplicating the data structure in the documentation.
If we were using the same specification format (Swagger) as Docker API, this would be a quick fix

**Comment by egeldenhuys** [comment]
Changes have been implemented in the spec.
Rendered spec can be viewed [here](https://tripleparity.github.io/docks-api/stack-api-spec.html)

**Pull request merged**